### PR TITLE
Rewrite in Ink

### LINKED ISSUE #34: Rewrite in Ink
https://github.com/vadimdemedes/ink

It's much better than Inquirer and results in more maintainable code.

We could use: https://github.com/maticzav/ink-autocomplete

Example of a CLI using Ink: https://github.com/sindresorhus/emoj/blob/master/ui.js

### PR DESCRIPTION
**Pull request by jopemachine**
Fixes #34.

It seems working now but I think I need to a little more time to look into more error handling logics and test codes and lint (WIP).
Thank you for all your efforts for maintaining awesome libs.

### CODE REVIEW

**Comment by jopemachine** [comment]
* This PR add `react`, `babel` to dependencies to rewrite in `ink`, and removes `inquirer` from dependencies.
* Some functions not having unit tests were tested myself directly in macos.
* Lots of existing logic in `interactive.js` are moved to `utils.js` to separate UI codes with process fetching, searching codes.
* I tried to maintain `fkill-cli`'s existing action as possible as I can, but some actions might be changed due to using `ink-select-input`, `ink-text-input`. If this couldn't be accepted, please let me know. I will try to make the components instead of using them.
* Some style (including text color) might have been changed a little bit.

**Comment by jopemachine** [comment]
It seems that this PR doesn't receive feedback yet.
May I ask if there are other things I need to do related to this issue?

**Comment by sindresorhus** [comment]
When searching, if you write a word, you should be able to press option+left-arrow to go back to the start, or option+right-arrow to go to the end. This worked before this PR.

**Comment by sindresorhus** [comment]
In case you are interested in making some improvements over the old version too:

- The "(Move up and down to reveal more choices)" text should only be shown on the first launch (of the interactive UI). Same with "(Use arrow keys or type to search)".
- You should be able to search for a port using the `:8080` syntax. I was sure this was already supported, but does not seem to work in the latest version.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -21,4 +21,5 @@ jobs:
         with:
           node-version: ${{ matrix.node-version }}
       - run: npm install
+      - run: npm run build
```
Comment (sindresorhus) [inline .github/workflows/main.yml:24]: Should be in the run script instead.

File: package.json [hunk 2]
```diff
@@ -41,26 +41,53 @@
 		"ps",
 		"proc"
 	],
+	"babel": {
+		"presets": [
+			"@babel/preset-react"
```
Comment (sindresorhus) [inline package.json:46]: Make sure that Babel does not transform ESM

File: interactive.js [hunk 3]
```diff
@@ -1,243 +1,184 @@
...
-import FuzzySearch from 'fuzzy-search';
+import hasAnsi from 'has-ansi';
+import {processExited, filterProcesses} from './utils.js';
+import {Dialog} from './dialog.js';
 
-const isWindows = process.platform === 'win32';
-const commandLineMargins = 4;
+const require = createRequire(import.meta.url);
+const TextInput = require('ink-text-input').default;
+const SelectInput = require('ink-select-input').default;
```
Comment (sindresorhus) [inline interactive.js:13]: Use `import`.

File: package.json [hunk 4]
```diff
@@ -12,17 +12,17 @@
...
 	},
 	"scripts": {
-		"test": "xo && ava"
+		"test": "xo && ava",
+		"build": "babel ./*.js --out-dir dist"
 	},
 	"files": [
-		"cli.js",
-		"interactive.js"
+		"./dist/*.js"
```
Comment (sindresorhus) [inline package.json:25]:
```suggestion
		"dist"
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: interactive.js (hunk 3 above)
Comment (jopemachine) [reply interactive.js:13]: I'm not sure the reason, but when using `import` on these ink libs, I've encountered the below error message.

```
Check the render method of `InteractiveUI`.
    at InteractiveUI (file:///Users/igyubong/Desktop/open-source/fkill-cli/distribution/interactive.js:20:3)
    at App (/Users/igyubong/Desktop/open-source/fkill-cli/node_modules/ink/build/components/App.js:42:9)

  ERROR  Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.

        Check the render method of `InteractiveUI`.

 - 
 - Check the render method of `InteractiveUI`.
 - createFiberFromTypeAndProps (node_modules/react-reconciler/cjs/react-reconciler.development.js:17777:21)
 - createFiberFromElement (node_modules/react-reconciler/cjs/react-reconciler.development.js:17805:15)
 - createChild (node_modules/react-reconciler/cjs/react-reconciler.development.js:4827:28)
 - reconcileChildrenArray (node_modules/react-reconciler/cjs/react-reconciler.development.js:5100:25)
 - reconcileChildFibers (node_modules/react-reconciler/cjs/react-reconciler.development.js:5506:14)
 - reconcileChildren (node_modules/react-reconciler/cjs/react-reconciler.development.js:8417:28)
 - updateHostComponent (node_modules/react-reconciler/cjs/react-reconciler.development.js:9059:3)
 - beginWork (node_modules/react-reconciler/cjs/react-reconciler.development.js:10507:14)
 - Object.invokeGuardedCallbackProd (node_modules/react-reconciler/cjs/react-reconciler.development.js:12101:10)
 - invokeGuardedCallback (node_modules/react-reconciler/cjs/react-reconciler.development.js:12292:31)
```

File: interactive.js [hunk 5]
```diff
@@ -1,243 +1,184 @@
...
-	}
-
-	await fkill(processes, {
-		force: true,
-		ignoreCase: true,
-	});
-};
+	const renderEmpty = () => (
+		<Box marginLeft={2}>
+			<Text color="#FF8000">No results...</Text>
```
[the remark this answers is not in the document]
Comment (jopemachine) [reply interactive.js]: I think possibly there might be no similar colors in [named colors](https://github.com/chalk/chalk/blob/main/source/index.d.ts#L10).
Could I change the color?
And I'm wondering if I need to use the named colors in other lines, too.

[the remark this answers is not in the document]
Comment (sindresorhus) [reply interactive.js]: `dimColor`

File: package.json (hunk 2 above)
Comment (jopemachine) [reply package.json:46]: > Make sure that Babel does not transform ESM

I was trying to use babel here only to use JSX.
(If possible, using [import-jsx](https://github.com/vadimdemedes/import-jsx) would be better, but [does not support ESM yet](https://github.com/vadimdemedes/import-jsx/issues/15))
Do we need to transform ESM to CJS or other tasks?

Comment (sindresorhus) [reply package.json:46]: > Do we need to transform ESM to CJS or other tasks?

Please re-read my comment. We do **not** want to transpile ESM to CJS. I think Babel does it by default.

https://babeljs.io/docs/en/options#targetsesmodules

Comment (jopemachine) [reply package.json:46]: > Please re-read my comment. We do **not** want to transpile ESM to CJS.

I'm sorry for the misunderstanding.

> I think Babel does it by default.

I used Babel only for using JSX here, so other transpiling configs not included.
You can see Babel is transpiling the source code to ESM itself under the `distribution` folder.

Comment (sindresorhus) [reply package.json:46]: > You can see Babel is transpiling the source code to ESM itself under the distribution folder.

I see that, but that contradicts their own docs:

> When no targets are specified: Babel will assume you are targeting the oldest browsers possible. 

- https://babeljs.io/docs/en/options#no-targets

Comment (jopemachine) [reply package.json:46]: > I see that, but that contradicts their own docs:

I posted a related question in the Babel discussions, and I got the below answer.

https://github.com/babel/babel/discussions/14533#discussioncomment-2700315

**Comment by sindresorhus** [comment]
@jopemachine Still interested in finishing this?

**Comment by jopemachine** [comment]
> @jopemachine Still interested in finishing this?

Yes, I'm definitely interested in this issue and other current issues of fkill.

So, there are some issues to look into.

---

- [ ] [Transpiling issue](https://github.com/sindresorhus/fkill-cli/pull/83#discussion_r862568130), I'm not sure about this.. Is this issue resolved now?

---

- [ ] [Require statement issue](https://github.com/sindresorhus/fkill-cli/pull/83#discussion_r862568229)
  
- [ ] When searching, if you write a word, you should be able to press option+left-arrow to go back to the start, or option+right-arrow to go to the end. This worked before this PR.

- [ ] Pressing Option+Delete when at the end of the search line, should clear the search text too.

: For implementing these features, I think updating `ink-text-input`, `ink-select-input` is required, but I don't think I can expect some reply at this point. I'm wondering what you think about how to handle this. What do you think removing these libs and reimplementing this logic?

---

- [ ] You should be able to search for a port using the :8080 syntax. I was sure this was already supported, but does not seem to work in the latest version.

: I think I can look into it when available.

**Comment by sindresorhus** [comment]
> https://github.com/sindresorhus/fkill-cli/pull/83#discussion_r862568130, I'm not sure about this.. Is this issue resolved now?

Check the output whether it's ESM or CJS. You may have set the `modules` option per the Babel comment.

**Comment by sindresorhus** [comment]
> https://github.com/sindresorhus/fkill-cli/pull/83#discussion_r862568229

Just add a TODO comment about fixing it when Ink properly supports ESM.

**Comment by sindresorhus** [comment]
> When searching, if you write a word, you should be able to press option+left-arrow to go back to the start, or option+right-arrow to go to the end. This worked before this PR.
  Pressing Option+Delete when at the end of the search line, should clear the search text too.

I think we can live without these for now. We'll just open new issues about fixing it at some point.

**Comment by sindresorhus** [comment]
> You should be able to search for a port using the :8080 syntax. I was sure this was already supported, but does not seem to work in the latest version.

Yes, that should ideally work. Not sure when it broke.

**Comment by jopemachine** [comment]
> Check the output whether it's ESM or CJS. You may have set the modules option per the Babel comment.

I'm not sure what is the Babel comment you mentioned, although the output is ESM. 

> Just add a TODO comment about fixing it when Ink properly supports ESM.

I think this is not required since `ink-text-input`, `ink-select-input` are not used now.

> I think we can live without these for now. We'll just open new issues about fixing it at some point.

I copy and pasted the above two repositories' source code and patched the `ink-text-input` for support delete, option key features in https://github.com/sindresorhus/fkill-cli/pull/83/commits/b157fd69c0cb59cdb3b1294d62328d255964cae4.

> Yes, that should ideally work. Not sure when it broke.

I'm not sure about the cause but I checked this feature worked normally in 5.2.0, but broke in 6.0.0, maybe inquirer 7's default behavior change? I added the feature since I don't see port-searching related code.

I'd appreciate your letting me know if there's anything else I could help.

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by sindresorhus** [comment]
Write `foo` in the prompt and then hold Option key while pressing Delete. The cursor moves backwards, but the text is not deleted. The expected behavior is that the whole text is erased after on press of Delete.

If you do the same, but hold Command instead Option, it adds a `u` character.

**Comment by sindresorhus** [comment]
Bump

**Comment by jopemachine** [comment]
@sindresorhus 
I am still interested in finishing this PR, but I am struggling with the key code testing logic on the cross-platform. 
Personally, I have been so busy with work that I apologize for the late response.
If there is anyone interested in this PR, I would really appreciate it if you could finish the work.
I will try to find time to complete this PR, but I am not sure exactly when that will be.

**Comment by sindresorhus** [comment]
No worries. But since there's no definite date of that happening, I prefer to close this for now to keep my PR queue clean. Feel free to submit a new pull request if/when you have time to finish this.

**Pull request closed without merging**